helper/logging/logging_http_transport.go

```diff
@@ -0,0 +1,252 @@
+			"error": err,
+		})
+	} else {
+		t.Debug(ctx, msg, fields)
+	}
+
+	// Invoke the wrapped RoundTrip now
+	res, err := t.transport.RoundTrip(req)
+	if err != nil {
+		return res, err
+	}
+
+	// Decompose the response bytes in a message (HTTP body) and fields (HTTP headers), then log it
+	msg, fields, err = decomposeResponseForLogging(res)
+	if err != nil {
+		t.Error(ctx, "Failed to parse response bytes for logging", map[string]interface{}{
+			"error": err,
+		})
+	} else {
+		t.Debug(ctx, msg, fields)
+	}
+
+	return res, nil
+}
+
+func (t *loggingHttpTransport) Debug(ctx context.Context, msg string, fields ...map[string]interface{}) {
+	if t.subsystem != "" {
+		tflog.SubsystemDebug(ctx, t.subsystem, msg, fields...)
+	} else {
+		tflog.Debug(ctx, msg, fields...)
+	}
+}
+
+func (t *loggingHttpTransport) Error(ctx context.Context, msg string, fields ...map[string]interface{}) {
+	if t.subsystem != "" {
+		tflog.SubsystemError(ctx, t.subsystem, msg, fields...)
+	} else {
+		tflog.Error(ctx, msg, fields...)
+	}
+}
+
+func decomposeRequestForLogging(req *http.Request) (string, map[string]interface{}, error) {
+	fields := make(map[string]interface{}, len(req.Header)+4)
+	fields[FieldHttpOperationType] = OperationHttpRequest
+
+	fields[FieldHttpRequestMethod] = req.Method
+	fields[FieldHttpRequestUri] = req.URL.RequestURI()
+	fields[FieldHttpRequestProtoVersion] = req.Proto
+
+	// Get the full body of the request, including headers appended by http.Transport:
+	// this is necessary because the http.Request at this stage doesn't contain
+	// all the headers that will be eventually sent.
+	// We rely on `httputil.DumpRequestOut` to obtain the actual bytes that will be sent out.
+	reqBytes, err := httputil.DumpRequestOut(req, true)
+	if err != nil {
+		return "", nil, err
+	}
+
+	// Create a reader around the request full body
+	reqReader := textproto.NewReader(bufio.NewReader(bytes.NewReader(reqBytes)))
+
+	err = readHeadersIntoFields(reqReader, fields)
+	if err != nil {
+		return "", nil, err
+	}
+
+	// Read the rest of the body content,
+	// into what will be the log message
+	return readRestToString(reqReader), fields, nil
+}
+
+func decomposeResponseForLogging(res *http.Response) (string, map[string]interface{}, error) {
+	fields := make(map[string]interface{}, len(res.Header)+4)
+	fields[FieldHttpOperationType] = OperationHttpResponse
+
+	fields[FieldHttpResponseProtoVersion] = res.Proto
+	fields[FieldHttpResponseStatusCode] = res.StatusCode
+	fields[FieldHttpResponseStatusReason] = res.Status
+
+	// Set the headers as fields to log
+	for k, v := range res.Header {
+		if len(v) == 1 {
+			fields[k] = v[0]
+		} else {
+			fields[k] = v
+		}
+	}
+
+	// Read the whole response body
+	resBody, err := io.ReadAll(res.Body)
+	if err != nil {
+		return "", nil, err
+	}
+
+	// Make a copy of the response body bytes, to be returned as log message
+	msgBytes := make([]byte, len(resBody))
+	copy(msgBytes, resBody)
```

Comment:
Drive-by note: Do we _need_ to copy the bytes here? We are not modifying them and they could potentially be "a lot" of memory. Having a reader pointing at the same bytes we originally read shouldn't have any effect in this case, I think.

Reply:
I don't know: we would have a shared piece of memory between the logging library, and the user of that response body.

And if we make modification to that shared portion of memory, we end up changing it under the hood for the user of the response body too.

Reply:
Downstream `io.Reader` consumers cannot modify the underlying bytes as far as I know and we control this implementation, but let's add a code comment explaining that choice then, e.g. "we're doing this in case its modified" 👍

Reply:
I'll give it a quick test: maybe I have over-thought this and it's way easier than I thought.

Reply:
Silly me: the string replacement we do during masking is going to generate a new string every time, so there is no chance of us messing up the same underlying byte buffer.

I can entirely remove this copying.